Alias

daniel-zid opened the pull request:
Simulations with alias (higher Nyquist zones) are now supported. A couple of bugs have been taken care of, such as:

- ADC resampling is now correctly implemented
- Noise trigger rate modules now use the correct window widths

cg-laser commented on NuRadioReco/modules/analogToDigitalConverter.py:
```diff
@@ -254,18 +282,63 @@ def get_digital_trace(self, station, det, channel,
         delayed_times = times + adc_time_delay
-        interpolate_trace = interp1d(times, trace, kind='quadratic',
-                                     fill_value='extrapolate')
+        interpolate_trace = interp1d(times, filtered_trace, kind='quadratic',
```
did you check that quadratic interpolation does not create artifacts, e.g., a quick significant over/under shooting of the values? Wouldn't a linear interpolation be better? 
An alternative (and likely saver approach) would be to upsample such that the downsampling factor is an integer. Then you can just take a subselection of the data points. E.g. if the downsampling factor is 2, just take very other sample.

daniel-zid replied:
Also, I've forgotten to comment something important. I cannot use a standard Fourier resampling method because it removes every component above half the Nyquist frequency, effectively destroying all the information in the higher Nyquist zones.

cg-laser replied:
yes, downsampling wouldn't work, but upsampling should be possible, shouldn't it? 
I guess my proposed method would work, but only for discrete time shift. E.g. if you upsample by a factor 10, you could choose the start sample from 0-9 and then take every 10th time bin. 
Maybe/probably the interplation method works better though. If you checked that the quadratic interpolation does not do weird stuff it might be the best. I just saw sometimes a weird behavior with higher order interpolations.

cozzyd replied:
I think upsampling followed by interpolation is the way to go. Interpolation causes phase issues at high frequencies and can add power at new frequencies for higher-order polynomials.  In my experience, upsampling by a factor of 4 and then linearly interpolating works decently well although you can of course do some numerical tests. 

If you're just trying to shift every sampled time by the same non-integer amount though, you can also use the Fourier Shift theorem (you just have to be careful around the boundaries since it's cyclic).   Basically a delay of \tau is in the time domain domain corresponds to multiplying by  exp(-j \omega \tau) in the Fourier domain (this is obvious if you think about the definition of group delay). But you'd want to either truncate the signal or add some zeroes (and then remove them) to avoid wraparound problems.  

Actually, unless you're losing a sample, you're going to have a boundary problem here too right?

daniel-zid replied:
> I think upsampling followed by interpolation is the way to go. Interpolation causes phase issues at high frequencies and can add power at new frequencies for higher-order polynomials. In my experience, upsampling by a factor of 4 and then linearly interpolating works decently well although you can of course do some numerical tests.
> 
> If you're just trying to shift every sampled time by the same non-integer amount though, you can also use the Fourier Shift theorem (you just have to be careful around the boundaries since it's cyclic). Basically a delay of \tau is in the time domain domain corresponds to multiplying by exp(-j \omega \tau) in the Fourier domain (this is obvious if you think about the definition of group delay). But you'd want to either truncate the signal or add some zeroes (and then remove them) to avoid wraparound problems.
> 

Yes, I think that would be better. I'll give it a try.
> Actually, unless you're losing a sample, you're going to have a boundary problem here too right?

That's right. But it's no big deal losing a sample. Or two or three. Voltages are long enough most of the time.

daniel-zid replied:
I don't think upsample is necessary. The input signals are already at 5 GHz, so I think that's more than enough for an interpolation. It would be nice to make sure that every input signal has a sampling rate of 5 GHz or greater, and if not, upsample the trace.

cozzyd replied:
agreed, as long as the sample rate is a few times higher than the highest frequency of interest then there's no need to upsample.

daniel-zid replied:
Now the delay is carried out using phases in Fourier space and the resulting trace is cropped. I've adjusted a bit the timing in the ADC module but ultimately it's not that relevant because it's just a global offset.

cg-laser commented on NuRadioReco/modules/io/eventWriter.py:
```diff
@@ -17,7 +17,7 @@ def get_header(evt):
             header['stations'][station.get_id()]['sim_station'] = {}
             header['stations'][station.get_id()]['sim_station'] = station.get_sim_station().get_parameters()
-    header['stations'][station.get_id()][stnp.station_time] = station.get_station_time()
+        header['stations'][station.get_id()][stnp.station_time] = station.get_station_time()
```
thanks for spotting this. It seems that it never worked for multiple stations

cg-laser commented on NuRadioReco/utilities/trace_utilities.py:
```diff
@@ -111,3 +112,116 @@ def get_electric_field_energy_fluence(electric_field_trace, times, signal_window
+    f = np.zeros_like(frequencies, dtype=np.complex)
+    mask = frequencies > 0
+    b, a = butter(order, passband, 'bandpass', analog=True)
+    w, h = freqs(b, a, frequencies[mask])
```
can you change the imports to `from scipy import signal.freqs` such that you need to call this function via `signal.freqs` or even better via `scipy.signal.freqs`. `freqs` is common variable name that will just override the function. So leaving it like this can be error prone.

daniel-zid replied:
Changed.

cg-laser commented on changelog.txt:
```diff
@@ -7,6 +7,7 @@ version 1.1.2 -
 new features:
 - Added the ability to force polarization to be only eTheta or ePhi in the reconstruction of the electric field using voltageToEfieldConverter.py
 - new detector which allows to change certain parameters of the detector description to study systematic uncertainties.
+- Simulations with alias (higher Nyquist zones) now supported
```
specify that this only applies to the phased array trigger simulator (I guess)

daniel-zid replied:
Changed.

daniel-zid commented:
> I had a brief look at the code. As far as I can say it looks good except for the inline comments I made.
> I didn't check the implementation of the phased array module thoroughly. I suggest that you ask someone who knows the details of this trigger to review the implementation (e.g. Cosmin).

Hi Christian. Thank you for the quick review. I will implement the requested changes as soon as I can. I'm going to ask Cosmin to review it as well.

cozzyd commented on NuRadioReco/utilities/trace_utilities.py:
```diff
@@ -111,3 +112,53 @@ def get_electric_field_energy_fluence(electric_field_trace, times, signal_window
+        raise ValueError(error_msg)
+
+    n_zeroes = int_factor - 1
+    zeroed_trace = []
```
This is fine for now but if performance matters later, it would be much better to make a zeroed trace first and fill in the non-zero samples (so you don't have so many needless memory allocations).

daniel-zid replied:
You're right. I can change that.

daniel-zid replied:
Changed.

cozzyd commented on NuRadioReco/utilities/trace_utilities.py:
```diff
@@ -111,3 +112,53 @@ def get_electric_field_energy_fluence(electric_field_trace, times, signal_window
+    upsampled_times = np.arange(0, len(zeroed_trace) * upsampled_delta_time, upsampled_delta_time)
+
+    cutoff = 1./int_factor
+    fir_coeffs = firwin(ntaps, cutoff)
```
Does the window choice matter here? Probably not very much but I think the way we've traditionally upsampled (zero-pad in frequency domain) is more akin to a rectangular window if I understand how firwin works.

daniel-zid replied:
I have no experience working with FIRs, so I've just used the default values. I can change it to `window=boxcar` if you like.

daniel-zid replied:
But yeah, I think that if we want the actual taps we need to use 'boxcar'. I didn't even realise. Let me try.

daniel-zid replied:
Well, waveforms look ok with the boxcar window. Might as well keep it.
![Figure_1](https://user-images.githubusercontent.com/45565159/80232236-dc52a500-8654-11ea-8635-e8e60eb75233.png)

cozzyd replied:
yeah, I I'm not sure what the best choice is!  I guess if the window is big enough it might not matter.

cozzyd commented on NuRadioReco/examples/AliasPhasedArray/Noise_trigger_rate/noise_trigger_rate.py:
```diff
@@ -0,0 +1,303 @@
+        if (upsampling_factor >= 2):
+            upsampling_factor = int(upsampling_factor)
+            upsampled_trace = upsampling_fir(digital_trace, adc_sampling_frequency,
+                                             int_factor=upsampling_factor, ntaps=2**7)
```
ntaps probably needs to be a function of the upsampling_factor? (or configurable)

cozzyd replied:
I guess I commented on the wrong commit, but same principle stands if you want this to be a little more general :).  Basically the reason is that the number of taps needed will scale with the upsampling factor because the taps are calculated on the upsampled waveform.

daniel-zid replied:
I've used 16 because that's what you told me. Is 16 fine for an upsampling of 4 or less? After that, how should I scale?

cozzyd replied:
If you've done tests and showed that it works well for that window length then that's probably fine!  However, it probably needs to scale with the upsampling factor (unless firwin automatically scales it or something).  Maybe it's a moot point if higher upsampling factors are never used?  If the window length is ever shorter than the number of zeroes stuffed between samples, though, that will cause weird problems.  Maybe upsampling_factor * 4 isn't a terrible rule of thumb?  You can also just sidestep it by making it an argument, since it could be an important parameter to vary in a study.

daniel-zid replied:
I have implemented your rule of thumb. The number of taps can be specified now using the detector configuration, in the same way as ADC sampling frequency or reference voltage.

cozzyd commented:
Quick readthrough looks sensible. I commented on a few things but they're probably not so important.

daniel-zid commented:
Alright, I've implemented the phased delay, now the number of taps can be included in the detector configuration file, and I've also changed the way to fill the zeroed trace. Let me know if you have any comments.

daniel-zid commented:
I found out that there was a bug in the noise trigger rate calculation. I'm rerunning some simulations to get the correct trigger rate. In principle this should not change much the Nyquist zone comparison, as the thresholds I was using were for tens of Hz instead of a few Hz as I expected, which means the efficiency of the three Nyquist zones were overestimated. And the second and third zones were underperforming anyway.

daniel-zid commented:
@cg-laser @cozzyd All the requested changes have been addressed. Could someone approve the pull request or tell me if there's anything left to change? Thanks!

cozzyd commented:
It could be that I don't properly know how to use the PR feature on Github, but I think @cg-laser has to approve that his requested changes were made?
